pkg/k8s/resource_ctors.go

```diff
@@ -51,6 +52,19 @@ func (def Config) Flags(flags *pflag.FlagSet) {
 	flags.String("k8s-service-proxy-name", def.K8sServiceProxyName, "Value of K8s service-proxy-name label for which Cilium handles the services (empty = all services without service.kubernetes.io/service-proxy-name label)")
 }
 
+// Dependencies for Cilium resources that may be used by Cilium Agent.
+// When CrdSyncPromise is provided, watchers of resources using this
+// will block until all CRDs used by the agent have been registered.
+// Agent will fail to start if Cilium Operator does not register all
+// the CRDs in time.
+type CiliumResourceParams struct {
+	cell.In
+
+	Lifecycle      cell.Lifecycle
+	ClientSet      client.Clientset
+	CrdSyncPromise promise.Promise[resource.CRDSync] `optional:"true"`
```

Comment:
Optional and nil-able are not the same.

The `optional` annotation instructs hive to not crash if `promise.Promise[resource.CRDSync]` isn't provided. It isn't needed here, even if the cell that provides it were to return `nil`. 

Actually, this would cause a silent error if the type is not provided instead of error-ing in CI and telling us about it.

Reply:
IMO this needs to be optional, as it may not be provided in other than agent hives. Other hives also use the k8s package, and I did not want to go and find all the different users and provide a dummy promise in those.

Reply:
So the background is that the k8s package is used by operator, and clustermech-apiserver, but it contains agent specific code for waiting for CRD sync, which is also used for this promise. Having the promise be optional makes it possible to those other binaries not have to wait for Agent's CRDs to be registered before observing a subset of those resources. We could make this non-optional, but then we'd have to make sure we provide a suitably different version on each different hive, implementations for which are not readily available.

I think we can alleviate the missing promise concern by not checking for `nil` in the daemon itself, so that it will crash if not provided.

Reply:
Made `CRDSync` promise required for the agent hive, optional for other hives.

Reply:
> IMO this needs to be optional, as it may not be provided in other than agent hives.

Understandable, I had the causality the other way around. 

> Made CRDSync promise required for the agent hive, optional for other hives.

Thanks, that should catch most of the issues I was worried about.